draco/util/random.py

```diff
@@ -0,0 +1,330 @@
+    mpi_random_seed
+"""
+import contextlib
+
+import numpy as np
+import zlib
+
+from caput import config
+from ..core import task
+
+
+_rng = None
+_default_bitgen = np.random.SFC64
+
+
+def default_rng():
+    """Returns an instance of the default random number generator to use.
+
+    This creates a randomly seeded generator using the fast SFC64 bit generator
+    underneath. This is only initialise on the first call, subsequent calls will
+    return the same Generator.
+
+    Returns
+    -------
+    rng : np.random.Generator
+    """
+
+    global _rng
+
+    if _rng is None:
+        _rng = np.random.Generator(_default_bitgen())
+
+    return _rng
+
+
+def complex_normal(
+    size=None, loc=0.0, scale=1.0, dtype=np.complex128, rng=None, out=None
+):
+    """Get a set of complex normal variables.
+
+        By default generate standard complex normal variables.
+
+        Parameters
+        ----------
+        size : tuple
+            Shape of the array of variables.
+        loc : float, optional
+            The mean of the complex output.
+        scale : float, optional
+            The standard deviation of the complex output.
+        dtype : {np.complex64, np.complex128}, optional
+            Output datatype.
+        rng : np.random.Generator, optional
+            Generator object to use.
+        out : np.ndarray[shape], optional
+            Array to place output directly into.
+    ,
+        Returns
+        -------
+        out : np.ndarray[shape]
+            Complex gaussian variates.
+    """
+
+    # Validate/set size argument
+    if size is None and out is None:
+        size = (1,)
+    elif out is not None and size is None:
+        size = out.shape
+    elif out is not None and size is not None and out.shape != size:
+        raise ValueError(
+            f"Shape of output array ({out.shape}) != size argument ({size}"
+        )
+
+    # Validate/set dtype argument
+    if dtype is None and out is None:
+        dtype = np.complex128
+    elif dtype is None and out is not None:
+        dtype = out.dtype
+    elif out is not None and dtype is not None and out.dtype != dtype:
+        raise ValueError(
+            f"Dtype of output array ({out.dtype}) != dtype argument ({dtype}"
+        )
+
+    if rng is None:
+        rng = default_rng()
+
+    _type_map = {
+        np.complex64: np.float32,
+        np.complex128: np.float64,
+    }
+
+    if dtype not in _type_map:
+        raise ValueError(
+            f"Only dtype must be complex64 or complex128. Got dtype={dtype}."
+        )
+
+    if out is None:
+        out = np.ndarray(size, dtype=dtype)
+
+    # Fill the complex array by creating a real type view of it
+    rtype = _type_map[dtype]
+    rsize = size[:-1] + (size[-1] * 2,)
+    rng.standard_normal(rsize, out=out.view(rtype))
+
+    # Use inplace ops for scaling and adding to avoid intermediate arrays
+    rscale = scale / 2 ** 0.5
+    out *= rscale
+
+    if loc != 0.0:
```

Comment:
I think this will cause an error if `loc` is an `np.ndarray`.
```suggestion
    if np.any(loc):
```